Bug 1909005: CamelCase and externalize K8s resources

rebeccaalpert opened the pull request:
CamelCased K8s resources in the nav, list pages, details pages, and kebab/action dropdowns. Externalized K8s resource labels in the public folder.

I didn't touch the K8s stuff in the Home tab because I'm not sure what's already been internationalized there - I may need to do that as a follow-on. I kept labels and labelPlurals in K8sKind because I'm not sure if they're still being used elsewhere.

Fixes https://issues.redhat.com/projects/CONSOLE/issues/CONSOLE-2440 and https://bugzilla.redhat.com/show_bug.cgi?id=1909005.

rebeccaalpert commented:
```
level=error msg=Cluster operator authentication Degraded is True with IngressStateEndpoints_MissingSubsets::OAuthServerConfigObservation_Error::OAuthServiceCheckEndpointAccessibleController_SyncError::OAuthServiceEndpointsCheckEndpointAccessibleController_SyncError::RouterCerts_NoRouterCertSecret: OAuthServiceEndpointsCheckEndpointAccessibleControllerDegraded: oauth service endpoints are not ready
level=error msg=OAuthServiceCheckEndpointAccessibleControllerDegraded: Get "https://172.30.245.4:443/healthz": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
level=error msg=RouterCertsDegraded: secret/v4-0-config-system-router-certs -n openshift-authentication: could not be retrieved: secret "v4-0-config-system-router-certs" not found
level=error msg=OAuthServerConfigObservationDegraded: secret "v4-0-config-system-router-certs" not found
level=error msg=IngressStateEndpointsDegraded: No subsets found for the endpoints of oauth-server
```

/test e2e-gcp-console

rebeccaalpert commented:
/test e2e-gcp-console

rebeccaalpert commented:
/test e2e-gcp-console

spadgett commented on frontend/public/models/index.ts:
```diff
@@ -3,7 +3,11 @@ import { K8sKind } from '../module/k8s';
   kind: 'Prometheus',
   label: 'Prometheus',
+  // t('k8s~Prometheus')
+  labelKey: 'k8s~Prometheus',
```
I was trying to think of a way to avoid having 4 properties in the model for labels, but I don't have a great suggestion :( It would be good to remove `label` and `labelPlural` from models that have a `labelKey`, but we'd need to sweep the code for everywhere we use `model.label`. That probably needs to be a follow on.

rebeccaalpert replied:
I'll add a follow-on task in Jira for this.

rebeccaalpert replied:
Added https://issues.redhat.com/browse/CONSOLE-2517.

rebeccaalpert commented on frontend/public/components/nav/admin-nav.tsx:
```diff
@@ -178,46 +178,46 @@ const AdminNav = () => {
 
-      <NavSection id="builds" title={t('nav~Builds')} required={FLAGS.OPENSHIFT}>
-        <ResourceNSLink id="buildconfigs" resource="buildconfigs" name={t('nav~Build Configs')} />
+      <NavSection id="buildconfigs" title={t('nav~Builds')} required={FLAGS.OPENSHIFT}>
```
Good catch! Thanks.

rebeccaalpert commented on frontend/public/components/utils/kebab.tsx:
```diff
@@ -79,17 +79,17 @@ const KebabItem_: React.FC<KebabItemProps & { isAllowed: boolean }> = ({
   const disabled = !isAllowed || option.isDisabled;
   const classes = classNames('pf-c-dropdown__menu-item', { 'pf-m-disabled': disabled });
-
+  const kind = option.labelKind && { kind: t(option.labelKind) };
```
I have a fallback on lines 89 and 92, but I can move it here to make it clearer.

rebeccaalpert replied:
I refactored to make this more clear.

rebeccaalpert commented on frontend/public/components/factory/list-page.jsx:
```diff
@@ -343,8 +343,8 @@ export const ListPage = withFallback((props) => {
   const ko = kindObj(kind);
-  const { label, labelPlural, namespaced, plural } = ko;
-  const title = props.title || labelPlural;
+  const { label, labelKey, labelPlural, labelPluralKey, namespaced, plural } = ko;
```
I tried doing i18next.t for these, but it didn't seem to work.

rebeccaalpert replied:
We do further down in the code. This just breaks out the values from the model. I tried doing i18next.t in the models originally, but I couldn't get it to work.

rebeccaalpert commented:
/retest

rhamilto reviewed (approved):
/lgtm

rhamilto commented:
/hold for approvals

/assign @yapei @ahardin-rh @sferich888

rebeccaalpert commented:
I'm moving this to a bug fix ticket so we can get @sg00dwin unblocked sooner. Removing hold and unassigning reviewers since we'll handle this under the bug fix process. (Paul Weil signed off on this.)

rebeccaalpert commented:
/bugzilla refresh

rebeccaalpert commented:
/bugzilla refresh

rebeccaalpert commented:
/bugzilla refresh

rhamilto commented:
/bugzilla refresh

yanpzhan commented on frontend/public/components/nav/admin-nav.tsx:
```diff
@@ -349,10 +349,10 @@ const AdminNav = () => {
         <ResourceNSLink
```
name={t('nav~Cluster Settings')}, if this nav needs to be CamelCased?

rebeccaalpert replied:
I left it un-CamelCased because we didn't have a model for it. Sam is out, but @rhamilto said this is ok to leave as-is. It's not a Kubernetes term.

yanpzhan commented on frontend/public/components/nav/admin-nav.tsx:
```diff
@@ -326,7 +326,7 @@ const AdminNav = () => {
           id="rolebindings"
           resource="rolebindings"
-          name={t('nav~Role Bindings')}
+          name={t('nav~RoleBindings')}
```
I found under one use/group page, the tab "Role Bindings" are not shown as "RoleBindings", need they be updated here?

rebeccaalpert replied:
Good catch! Thanks @yanpzhan.

rebeccaalpert replied:
Addressed this (and a couple more I found on a second pass). Thanks again!

rhamilto commented:
New changes LGTM.  Please squash.

rebeccaalpert commented:
Squashed.

rhamilto commented:
/lgtm

rhamilto commented:
@rebeccaalpert, needs rebasing.

rhamilto commented:
/lgtm

rhamilto commented:
/lgtm

rhamilto commented:
/retest

yanpzhan commented:
/label qe-approved

spadgett commented:
/hold for approvals

spadgett commented:
/hold cancel

spadgett commented:
@rebeccaalpert FYI, this needs rebase

rebeccaalpert commented:
Thanks @spadgett; rebased!

spadgett commented:
@rhamilto Can you give a final review since this has been rebased?

rhamilto reviewed (approved):
/lgtm

rhamilto commented:
I assume the namespace consolidation will result in de-duplication, so I went ahead and tagged.

spadgett commented:
/lgtm cancel

The following test is failing each run:
> Performance test: downloads new bundle for service-account

spadgett commented on frontend/integration-tests/tests/performance.scenario.ts:
```diff
@@ -7,28 +7,28 @@ import * as crudView from '../views/crud.view';
+  .set('build-config', { section: 'Builds', name: 'BuildConfigs' })
+  .set('image-stream', { section: 'Builds', name: 'ImageStreams' })
   .set('node', { section: 'Compute', name: 'Nodes' })
   .set('service-account', { section: 'User Management', name: 'Service Accounts' })
```
It looks like this line needs to be updated, too.

rebeccaalpert replied:
Good catch, thank you!

rebeccaalpert replied:
All set @spadgett.

rhamilto commented:
@rebeccaalpert, I see you've force pushed a couple times.  Does this need another review?  Or are you simply rebasing?

rebeccaalpert commented:
@rhamilto - Can I ping you when it needs another review? I'm trying to get the e2e test to pass.

rebeccaalpert commented:
/retest

TheRealJon commented on frontend/public/components/utils/kebab.tsx:
```diff
@@ -256,7 +256,7 @@ const kebabFactory: KebabFactory = {
     // t('details-page~Delete {{kind}}', {kind: kind.label})
     labelKey: 'details-page~Delete {{kind}}',
-    labelKind: { kind: kind.label },
+    labelKind: { kind: i18next.t(kind.labelKey) },
```
We're using an internationalized string as a parameter in another internationalized string. Wondering if this might cause issues. For example, if the non-parameterized part of the string is reactive and the parameter isn't, the user might select a new language and only part of the string would change.

rebeccaalpert replied:
You can nest them. We do this elsewhere in the application as well.

TheRealJon reviewed (approved):
A few small comments.

rebeccaalpert commented:
Spent some time debugging e2e tests; made some changes. Hoping to see CI test failure reductions, but will continue investigating locally as well.

Addressed PR feedback and consolidated namespaces.

spadgett commented:
/retest

rebeccaalpert commented:
Rebased.

rebeccaalpert commented:
```
level=error
level=error msg=Error: Request "Create IAM Members roles/storage.admin serviceAccount:ci-op-895lpiks-75d12-gd67j-w@openshift-gce-devel-ci.iam.gserviceaccount.com for \"project \\\"openshift-gce-devel-ci\\\"\"" returned error: Error applying IAM policy for project "openshift-gce-devel-ci": Error setting IAM policy for project "openshift-gce-devel-ci": googleapi: Error 400: Service account ci-op-gmqict-openshift-i-4r84q@openshift-gce-devel-ci.iam.gserviceaccount.com does not exist., badRequest
level=error
level=error msg=  on ../tmp/openshift-install-654106175/iam/main.tf line 11, in resource "google_project_iam_member" "worker-storage-admin":
level=error msg=  11: resource "google_project_iam_member" "worker-storage-admin" {
level=error
level=error
level=fatal msg=failed to fetch Cluster: failed to generate asset "Cluster": failed to create cluster: failed to apply Terraform: failed to complete the change
error: failed to execute wrapped command: exit status 1
`

/test e2e-gcp-console

rebeccaalpert commented:
operator-install-single-namespace.spec.ts is failing on CI but runs well locally. Maybe a flake? /test e2e-gcp-console

rhamilto commented:
> operator-install-single-namespace.spec.ts is failing on CI but runs well locally. Maybe a flake? /test e2e-gcp-console

@TheRealJon I think we may have an issue with the create operator form as this is showing up quite a bit where the form is just a blank white page.

<img width="1281" alt="Screen Shot 2021-01-22 at 2 55 05 PM" src="https://user-images.githubusercontent.com/895728/105538853-07d8ed80-5cc2-11eb-83c8-8ae7136285f9.png">

rebeccaalpert commented:
Rebased and i18ned breadcrumb Dave added.

rhamilto commented:
/lgtm

rhamilto commented:
/hold cancel

heads up, @sg00dwin.
